Objective-C/CBLCollection.mm

```diff
@@ -976,4 +831,171 @@ - (nullable CBLQueryIndex*) indexWithName: (nonnull NSString*)name
+    CBL_LOCK(_mutex) {
+        if (!_colObs || !_c4col)
+            return;
+        
+        const uint32_t kMaxChanges = 100u;
+        C4DatabaseChange changes[kMaxChanges];
+        bool external = false;
+        C4CollectionObservation obs = {};
+        NSMutableArray* docIDs = [NSMutableArray new];
+        do {
+            // Read changes in batches of kMaxChanges:
+            obs = c4dbobs_getChanges(_colObs, changes, kMaxChanges);
+            if (obs.numChanges == 0 || external != obs.external || docIDs.count > 1000) {
+                if(docIDs.count > 0) {
+                    CBLCollectionChange* change = [[CBLCollectionChange alloc] initWithCollection: self
+                                                                                      documentIDs: docIDs
+                                                                                       isExternal: external];
+                    [_colChangeNotifier postChange: change];
+                    docIDs = [NSMutableArray new];
+                }
+            }
+            
+            external = obs.external;
+            for(uint32_t i = 0; i < obs.numChanges; i++) {
+                NSString *docID =slice2string(changes[i].docID);
+                [docIDs addObject: docID];
+            }
+            c4dbobs_releaseChanges(changes, obs.numChanges);
+        } while(obs.numChanges > 0);
+    }
+}
+
+- (void) removeToken: (id)token {
+    CBL_LOCK(_mutex) {
+        CBLChangeListenerToken* t = (CBLChangeListenerToken*)token;
+        if (t.context)
+            [self removeDocumentChangeListenerWithToken: token];
+        else {
+            if ([_colChangeNotifier removeChangeListenerWithToken: token] == 0) {
+                c4dbobs_free(_colObs);
+                _colObs = nil;
+                _colChangeNotifier = nil;
+            }
+        }
+    }
+}
+
+- (void) removeDocumentChangeListenerWithToken: (CBLChangeListenerToken*)token {
+    CBL_LOCK(_mutex) {
+        NSString* documentID = (NSString*)token.context;
+        CBLDocumentChangeNotifier* notifier = _docChangeNotifiers[documentID];
+        if (notifier && [notifier removeChangeListenerWithToken: token] == 0) {
+            [notifier stop];
+            [_docChangeNotifiers removeObjectForKey:documentID];
+        }
+    }
+}
+
+- (void) freeC4Observer {
+    c4dbobs_free(_colObs);
+    _colObs = nullptr;
+    _colChangeNotifier = nil;
+
+    [_docChangeNotifiers.allValues makeObjectsPerformSelector: @selector(stop)];
+    _docChangeNotifiers = nil;
+}
+
+- (nullable NSArray*) _indexesInfo: (NSError**)error {
+    CBL_LOCK(_mutex) {
+        if (![self checkIsValid: error])
+            return nil;
+        
+        C4Error err = {};
+        C4SliceResult res = c4coll_getIndexesInfo(_c4col, &err);
+        if (err.code != 0){
+            convertError(err, error);
+            return nil;
+        }
+        
+        FLDoc doc = FLDoc_FromResultData(res, kFLTrusted, nullptr, nullslice);
```

Comment:
Wondering if FLDoc requires the `res` to be a dictionary only or not (I don't have the info).